### DISCUSSION TITLE
How do I hit Gui+Shift+S?

### DISCUSSION THREAD

**Discussion by abandon42** [manna-harbour/miryoku#98] [Q&A] [answered]
Stupid question, I'm new.

**Comment by dcolestock** [comment, marked as answer]
First, on the opposite hand from the s, press both the home row index finger (shift) and the home row pinky finger (gui) then press `s` with the other hand and then release them.

**Reply by abandon42** [reply]
ok, my brain just melted. Thank you

**Comment by manna-harbour** [comment]
As @dcolestock already mentioned, use the opposite hand for the mods.

A User Guide will be added, but at the moment the only documentation is the [Reference Manual](https://github.com/manna-harbour/miryoku/tree/master/docs/reference), which takes some interpretation.

The basic idea for mods and / or layers is to hold mods on fingers and layer change on thumb simultaneously with one hand, and use the other to reach for the key on the layer and / or to be combined with the mods. That way, the hand that is holding doesn't move from home, and the hand moving doesn't hold. So both hands can be relaxed, overall movement is minimised, hand contortions are eliminated, and any combination of mods can be used with any key.

**Comment by abandon42** [comment]
Thanks! I have most of it now. New to QMK and home row mods.

Now just have to fork it and build a custom layout for unused keyson the
iris6.

A but unclear on:

   - Auto Shift for numbers and symbols.
   - Secondary left-hand layer has shifted symbols in the same locations to
   reduce chording when using *mods with shifted symbols. *
   - The additional feature layer. Don't understand the use cases. Qwerty
   would be helpful if someone has to use the keyboard, and would ideally use
   the default base layer for the keyboard with some way back?

Thanks for all your work on this! The thumb backspace is amazing for
someone learning colemak. Now I just have to figure out what I want all the
extra keys to do.

(Ps, iris needed the console disabled for filesize reasons, and the dev
fork would segfault on compile. I don't figure configurator can handle
this? Via is not supported due to layer count?)

**Reply by manna-harbour** [reply]
See https://precondition.github.io/home-row-mods.

To map unused keys see https://github.com/manna-harbour/miryoku/discussions/85. They aren't needed, but you might like to use them as you would a macro pad. Don't feel compelled to use all the keys just because they exist. The keys that are used were chosen purely for ergonomics, not to fit any particular keyboard. It was originally developed on an ergodox, where more than half the keys are unused.

Shift or auto shift can be used on Num to reach shifted symbols. Auto shift is activated by holding the key, see https://docs.qmk.fm/#/feature_auto_shift. The Sym layer is a convenience for those who prefer the thumb movement to holding the home thumb plus shift. That might be more important if you're also holding another mod, and less important if you're already on Num. Sym also has matching parens on the thumbs for convenience.

The Additional Features section describes the Extra and Tap layers as well as layer lock. Extra isn't for any particular purpose, but can be used if you're in the process of learning a new alpha arrangement, for remapping in the OS, or for use by others. You can switch between Base and Extra using the keys indicated.

For backspace, note that backspace on Nav is a basic keycode and will auto repeat. But it's generally better to combine with control instead (or as well).

Console should really be disabled at the keyboard level for all keyboards by default.

QMK `develop` is unstable. The issue seems to come from https://github.com/qmk/qmk_firmware/pull/17229. And related to earlier changes in https://github.com/qmk/qmk_firmware/pull/8311 (which aren't breaking at that is in `master`). You could try building the default keymap from `develop` and if that fails, provide feedback to QMK.  I'll have to look into it though, as they might be making the assumption that the `keymaps` array is in `keymap.c`, (see https://github.com/manna-harbour/miryoku_qmk/runs/7073304591?check_suite_focus=true#step:7:24) which isn't the case for Miryoku...

The configurator lacks support for advanced features. You can enable Via / Vial and just increase the layer limit. You might need to reduce the firmware size in other areas though, see https://docs.qmk.fm/#/squeezing_avr. But to customise layers also see the customisation docs linked above.

**Reply by abandon42** [reply]
Thanks, I'll have to try the new customization framework. Having left pinky control for one handed shortcuts is helpful! I'm also using numbers on the top row, enough to learn with colemak. I just modified it directly, it mostly works except for the upper left KB_ESC for some reason. Will report back.

I understand autoshift now.

For some reason, I can't get the additional features layer to work. I think it's right thumb held with upper right pinky? Can't get them to do anything.

And no idea what's up with the button layer. How is it activated?

**Reply by manna-harbour** [reply]
For Additional Features, hold any thumb then use the keys shown here on the same hand as the held thumb: https://github.com/manna-harbour/miryoku/tree/master/docs/reference#additional-features. But you also need to be using the `miryoku` branch https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#miryoku-qmk-development-branch as Tap, Extra, and layer lock are not in the version of Miryoku currently in QMK `master`. Right thumb with right top row pinkie will activate the bootloader.

Button https://github.com/manna-harbour/miryoku/tree/master/docs/reference#button is primarily for use with pointing devices, particularly integrated or buttonless. But can also be used for one hand clipboard.

**Reply by abandon42** [reply]
Ah, that's why I've been struggling with additional features! I tried switching to the miryoku origin already, but that one was segfaulting for the iris6 at least. 

Is there any way to activate the button layer with an external usb mouse?

**Reply by manna-harbour** [reply]
iris should build directly from `miryoku` https://github.com/manna-harbour/miryoku_qmk/actions/runs/2569309487, or when combined with upstream `master` with console disabled https://github.com/manna-harbour/miryoku_qmk/actions/runs/2569356446 . Combining with upstream `develop` is currently broken as discussed earlier https://github.com/manna-harbour/miryoku_qmk/actions/runs/2569368341.

**Reply by manna-harbour** [reply]
This will treat an external pointing device as if it was internal, including auto button layer activation: https://github.com/manna-harbour/qmk_firmware/issues/40. It needs to be rewritten, and support for external pointing devices is currently Linux only, but if you subscribe to that issue you'll be notified of updates.

**Comment by abandon42** [comment]
![image](https://user-images.githubusercontent.com/13699448/176273492-a7d28eeb-0368-44d4-a39f-e0bd9c3f63f2.png)

I'm definitely doing something wrong with custom config. 

Got Via working by manually setting VIA_ENABLE=yes. Down to 644 bytes free! I now have a JSON of Miryoku for Iris6a, only took me 5 hours to get here ;)  Any interest in a community resource of keymaps? Seems it would help people to try Miryoku if they haven't built QMK extensively.

**Reply by abandon42** [reply]
No, I was mistaken. VIA detects the keyboard, but layers, mouse are broken.

**Reply by manna-harbour** [reply]
That mapping looks ok in general. You'll need the define for `XXX` as they are still used in the mapping. Also see above regarding the `miryoku` branch.

For via you'll need to increase the layer limit. Also, you might like to try vial instead.

**Reply by abandon42** [reply]
Is custom_config read instead of config when it's present? Thought the define would've been captured there.

**Reply by manna-harbour** [reply]
No, you're correct. I was thinking of Miryoku ZMK, where the mappings are handled slightly differently.

**Reply by abandon42** [reply]
Any recommendations on more features I can disable to up the layer count? I'm maxing out at 6 with console off and via enabled. 

I was able to get a default build of Vial, but not miryoku yet; above layer challenge applies, and I think more needs porting.

**Reply by manna-harbour** [reply]
In general see https://docs.qmk.fm/#/squeezing_avr and also check what else is enabled by default at the keyboard level for that keyboard. For Miryoku specifically, candidates for removal, depending on your needs, are the Extra, Tap, Button, Mouse, Fun, and Sym layers (https://github.com/manna-harbour/miryoku/discussions/85), and the `AUTO_SHIFT`, `MOUSEKEY`, and `EXTRAKEY` features.

**Comment by abandon42** [comment]
These are defined already by Iris:

BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output


I tried the following:

#define LAYER_STATE_8BIT
#define NO_ACTION_ONESHOT
#undef LOCKING_SUPPORT_ENABLE
#undef LOCKING_RESYNC_ENABLE

CONSOLE_ENABLE = no
SPACE_CADET_ENABLE = no
GRAVE_ESC_ENABLE = no
MAGIC_ENABLE = no
AVR_USE_MINIMAL_PRINTF = yes


As well as disabling all the RGB anims. Haven't tried disabling the entire
subsystem.

If DYNAMIC_KEYMAP_LAYER_COUNT exceeds *7, *I get :"Dynamic keymaps are
configured to use more EEPROM than is available." Controller only has 1kb.

Otherwise at 7 firmware size is fine - * The firmware size is fine -
21732/28672 (75%, 6940 bytes free)

Guess I could sacrifice the button layer. Maybe another time. I'm not sure
what'll happen to EEPROM when I start enabling features I actually want in
VIA/Vial, like encoder control.

**Reply by manna-harbour** [reply]
That seems ok. You might need to check with Via or QMK support if it doesn't build with more than 7 layers.

**Reply by abandon42** [reply]
Thanks.

I can't build miryoku using the miryoku repo. Is there something that prevents it from building in the windows qmk env?

Here's what I did:

```
git clone https://github.com/manna-harbour/miryoku_qmk.git
cd miryoku_qmk
make submodule
make keebio/iris/rev6:manna-harbour_miryoku CONSOLE_ENABLE=no
```
>Segmentation fault (core dumped)

This fails on rev3:default as well.

I had the same issue using your directions on my qmk head on a previously functioning directory - 
 
```
git remote add miryoku_qmk git@github.com:manna-harbour/miryoku_qmk.git # if using SSH
git remote add miryoku_qmk https://github.com/manna-harbour/miryoku_qmk.git # if using HTTPS
git fetch miryoku_qmk
git checkout --track miryoku_qmk/miryoku
```

My QMK ver goes from QMK Firmware 0.17.5 to QMK Firmware 0.14.23 with the above directions.

**Reply by manna-harbour** [reply]
There must be an issue with your MSYS build environment, as `miryoku` still builds via GitHub Actions, and it fails for the default keymap when you build locally.  It might be that the old version of QMK in the `miryoku` branch is no longer compatible with the new version of the MSYS build environment you have set up for building with upstream `master`. I can't really help with windows, but I'll be rebasing the `miryoku` branch this week, so it that's the issue, that should take care of it.

To build directly from current `miryoku` now, instead of MSYS you could build with WSL or docker, see the QMK docs, or GitHub Actions, see https://github.com/manna-harbour/miryoku/discussions/98#discussioncomment-3052544 and https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#workflow-builds.

Alternatively, you could merge or rebase `miryoku` on upstream `master`. If you do that, drop the `[local]` commits first to avoid merge conflicts. Or as a PR was recently merged into `master`, you could cherry pick the commits since then, see https://github.com/manna-harbour/miryoku_qmk/compare/miryoku-20220227...miryoku.

**Reply by manna-harbour** [reply]
> Alternatively, you could merge or rebase `miryoku` on upstream `master`.

Here is a temporary rebased branch: https://github.com/manna-harbour/miryoku_qmk/tree/miryoku-rebase-20220713-tmp. You should be able to build from that directly using the same build environment as with upstream `master`. That branch will replace the current `miryoku` branch soon.

**Reply by abandon42** [reply]
Build successful, thanks!

**Reply by abandon42** [reply]
> That seems ok. You might need to check with Via or QMK support if it doesn't build with more than 7 layers.

I'll have to follow up with them. In the meantime, I checked the code for the EEPROM requirements:
`DYNAMIC_KEYMAP_ENCODER_EEPROM_ADDR + (DYNAMIC_KEYMAP_LAYER_COUNT * NUM_ENCODERS * 2 * 2)`

Since the Iris has 2 encoders, my understanding is that I run out of eeprom past 7 layers, but I'm not sure if encoders means encoders here, because disabling encoders in rules.mk doesn't fix it. Also I'd like to use my encoders.

So I've run away screaming for now, and will just program it directly. Vial uses even more space.

**Reply by abandon42** [reply]
Great, so in the new branch, I got custom_config working. Very nice design, my first experience with C, I barely understand what's going on.

For:

> As mappings need to be specific to the keyboard, everything should be wrapped in #if defined / #endif conditional on the automatically defined KEYBOARD_ plus keyboard name macro.

Would it be KEYBOARD_iris_rev6? 'iris' didn't work. 

I also got encoder functionality changing with layers. The code is in the keyboard's folder, and I couldn't figure out how to reference the miryoku layer enum, so I just used integers. Feels a bit sloppy overall, and would appreciate any tips!

Thanks again
